python/lsst/daf/ingest/ingestCatalogTask.py

```diff
@@ -0,0 +1,576 @@
+            keys.append((item.key, fieldFormatters[item.field.getTypeString()]))
+            columnNames.append(self._columnName(item.field.getName()))
+        sqlPrefix += ",".join(columnNames)
+        sqlPrefix += ") VALUES "
+        pos = 0
+        while pos < len(cat):
+            sql = sqlPrefix
+            initialPos = pos
+            maxValueLen = maxQueryLen - len(sql)
+            while pos < len(cat):
+                row = cat[pos]
+                value = "("
+                value += ",".join([f.formatValue(row.get(k)) for (k, f) in keys])
+                value += "),"
+                maxValueLen -= len(value)
+                if maxValueLen < 0:
+                    break
+                else:
+                    sql += value
+                    pos += 1
+            if pos == initialPos:
+                # Have not made progress
+                raise RuntimeError("Single row is too large to insert")
+            self._executeSql(conn, sql[:-1])
+        conn.commit()
+
+    def _columnName(self, fieldName):
+        """Return the SQL column name for a given afw table field.
+        """
+        name = canonicalizeFieldName(fieldName)
+        return self.config.remap.get(name, name)
+
+    def _columnDef(self, field):
+        """Return the SQL column definition for a given afw table field.
+        """
+        sqlType = fieldFormatters[field.getTypeString()].sqlType(field)
+        return self._columnName(field.getName()) + " " + sqlType
+
+    def _createTable(self, conn, tableName, schema):
+        """Create a table using column definitions based on the provided afw
+        table schema, adding in any extra columns specified in the config.  If
+        one exists, the unique id column is given a key.
+        """
+        fields = [item.field for item in self._schemaItems(schema)]
+        names = [self._columnName(field.getName()) for field in fields]
+        if len(names) != len(set(n.lower() for n in names)):
+            raise RuntimeError(
+                "Schema contains columns that differ only by non-word characters "
+                "and/or case. In the database, these cannot be distinguished and "
+                "hence result in column name duplicates. Use the remap "
+                "configuration parameter to resolve this ambiguity.")
+        sql = "CREATE TABLE IF NOT EXISTS {} (\n\t".format(tableName)
+        sql += ",\n\t".join(self._columnDef(field) for field in fields)
+        if self.config.extraColumns:
+            sql += ",\n\t" + self.config.extraColumns
+        if self.config.idColumnName:
+            if self.config.idColumnName in names:
+                sql += ",\n\tUNIQUE({})".format(self.config.idColumnName)
+            else:
+                self.log.warn(
+                    "No field matches the configured unique ID column name "
+                    "({})".format(self.config.idColumnName))
+        sql += "\n)"
+        self._executeSql(conn, sql)
```

Comment:
Yes, we'll throw and it'll just propagate. That'll either be because of some runtime MySQL problem, in which case that's the right thing to do, or because I'm producing some invalid SQL. The only way the latter can happen (modulo bugs in my code) is if the user specifies bogus or duplicate columns in `config.extraColumns`. Hopefully MySQL comes back with an understandable error message in that case. If it doesn't, producing a better message would probably involve parsing SQL column specs (maybe one could get by with just interpreting database specific error codes, or changing the way extra columns are specified). I don't really want to go down any of these paths until I hear users complaining.